### PR TITLE
CHANGE (CodeAnalyzer): @W-13594221@: Added CPD and pmd-custom checks to smoke tests.

### PR DESCRIPTION
**Pull request by jfeingold35**

### CODE REVIEW

File: smoke-tests/SmokeTestGenerator.js [hunk 1]
```diff
@@ -45,43 +45,67 @@ function buildPath(pathSegs, delim) {
...
+		`${exeName} scanner:run --format junit --target ${buildPath([...projectsPath, 'ts', 'src', 'simpleYetWrong.ts'], delim)} --tsconfig ${buildPath([...projectsPath, 'tsconfig.json'], delim)} --outfile ${buildPath([...resultsPath, 'eslint-typescript.xml'], delim)}`,
+		// Run such that RetireJS is invoked.
 		`echo "==== Run RetireJS against a folder ===="`,
-		`${exeName} scanner:run --format junit --engine retire-js --target ${buildPath([...projectsPath, 'dep-test-app', 'folder-a'], delim)} --outfile ${buildPath([...resultsPath, 'run3.xml'], delim)}`,
+		`${exeName} scanner:run --format junit --engine retire-js --target ${buildPath([...projectsPath, 'dep-test-app', 'folder-a'], delim)} --outfile ${buildPath([...resultsPath, 'retire-js.xml'], delim)}`,
+		// Run such that CPD is invoked.
+		`echo "==== Run CPD against a folder ===="`,
+		`${exeName} scanner:run --format junit --engine cpd --target ${buildPath([...projectsPath, 'cpd-test-app', 'src', 'classes'], delim)} --outfile ${buildPath([...resultsPath, 'cpd.xml'])}`,
```
Comment (jfeingold35) [inline smoke-tests/SmokeTestGenerator.js:83]: This part is net-new.

File: smoke-tests/SmokeTestGenerator.js [hunk 2]
```diff
@@ -45,43 +45,67 @@ function buildPath(pathSegs, delim) {
...
+		// Run such that PMD is invoked with a custom config.
+		`echo "==== Run PMD with custom config via --pmdconfig flag ===="`,
+		`${exeName} scanner:run --format junit --engine pmd --target ${buildPath([...projectsPath, 'app', 'force-app'], delim)} --pmdconfig ${buildPath(customConfigPath, delim)} --outfile ${buildPath([...resultsPath, 'pmd-customconfig.xml'], delim)}`,
```
Comment (jfeingold35) [inline smoke-tests/SmokeTestGenerator.js:86]: This part is also net-new.

File: smoke-tests/SmokeTestGenerator.js [hunk 3]
```diff
@@ -45,43 +45,67 @@ function buildPath(pathSegs, delim) {
...
 	const projectsPath = ['test', 'code-fixtures', 'projects'];
 	const customRulePath = ['test', 'test-jars', 'apex', 'testjar1.jar'];
+	const customConfigPath = ['test', 'test-xml', 'apex', 'smoke-config.xml'];
 	const resultsPath = ['smoke-test-results'];
 
 	const exeName = isBash ? '$EXE_NAME' : '%EXE_NAME%';
 
 	// Declare an array with all of the commands we intend to execute.
 	const commands = [
+		// Log a header.
```
Comment (jfeingold35) [inline smoke-tests/SmokeTestGenerator.js:55]: Added commenting between lines in order to more clearly delineate different stages.

File: smoke-tests/SmokeTestGenerator.js [hunk 4]
```diff
@@ -45,43 +45,67 @@ function buildPath(pathSegs, delim) {
...
+		// Describe a real rule.
 		`echo "==== Describe a real rule ===="`,
 		`${exeName} scanner:rule:describe -n EmptyCatchBlock`,
+		// Describe a non-existent rule.
 		`echo "==== Describe a non-existent rule ===="`,
 		`${exeName} scanner:rule:describe -n NotAnActualRule`,
+		// Run such that PMD and ESLint are invoked.
 		`echo "==== Run rules against force-app, which should hit PMD and ESLint engines ===="`,
-		`${exeName} scanner:run --format junit --target ${buildPath([...projectsPath, 'app', 'force-app'], delim)} --outfile ${buildPath([...resultsPath, 'run1.xml'], delim)}`,
+		`${exeName} scanner:run --format junit --target ${buildPath([...projectsPath, 'app', 'force-app'], delim)} --outfile ${buildPath([...resultsPath, 'pmd-eslint.xml'], delim)}`,
```
Comment (jfeingold35) [inline smoke-tests/SmokeTestGenerator.js:74]: Changed all of the `--outfile` arguments to have more descriptive names instead of `runX.xml`.

Comment (rmohan20) [reply smoke-tests/SmokeTestGenerator.js:74]: Great idea to put additional information in the file name. Could you also include the word "results" somewhere? Would help someone who has less experience with our smoke tests.

**Review by rmohan20 — APPROVED** [review]

**Pull request merged**